diff --git a/sherlodoc/index/load_doc.ml b/sherlodoc/index/load_doc.ml
index 7f78397..c4980a5 100644
--- a/sherlodoc/index/load_doc.ml
+++ b/sherlodoc/index/load_doc.ml
@@ -81,7 +81,7 @@ let searchable_type_of_constructor args res =
   match args with
   | TypeDecl.Constructor.Tuple args -> begin
       match args with
-      | _ :: _ :: _ -> TypeExpr.(Arrow (None, Tuple args, res))
+      | _ :: _ :: _ -> TypeExpr.(Arrow (None, Tuple (List.map (fun a -> (None, a)) args), res))
       | [ arg ] -> TypeExpr.(Arrow (None, arg, res))
       | _ -> res
     end
@@ -154,6 +154,7 @@ let rec categorize id =
   | `AssetFile _ | `SourceLocationMod _ | `SourceLocation _ | `SourcePage _
   | `SourceLocationInternal _ ->
       `ignore (* unclear what to do with those *)
+  | _ -> `ignore
 
 let categorize Odoc_index.Entry.{ id; _ } =
   match id.iv with
diff --git a/sherlodoc/index/type_cache.ml b/sherlodoc/index/type_cache.ml
index 3d85846..186e581 100644
--- a/sherlodoc/index/type_cache.ml
+++ b/sherlodoc/index/type_cache.ml
@@ -19,5 +19,5 @@ let rec of_odoc ~cache otyp =
   | Arrow (_lbl, left, right) -> cache (Arrow (of_odoc ~cache left, of_odoc ~cache right))
   | Constr (name, args) ->
       cache (Constr (Typename.to_string name, List.map (of_odoc ~cache) args))
-  | Tuple li -> cache (Tuple (List.map (of_odoc ~cache) li))
+  | Tuple li -> cache (Tuple (List.map (fun (_lbl, ty) -> of_odoc ~cache ty) li))
   | _ -> Unhandled
diff --git a/sherlodoc/jsoo/tyxml.ml b/sherlodoc/jsoo/tyxml.ml
index 3ebabd7..bc8b67c 100644
--- a/sherlodoc/jsoo/tyxml.ml
+++ b/sherlodoc/jsoo/tyxml.ml
@@ -22,7 +22,6 @@ end = struct
 
   let add_escape_string buf s =
     (* https://discuss.ocaml.org/t/html-encoding-of-string/4289/4 *)
-    let add = Buffer.add_string buf in
     let len = String.length s in
     let max_idx = len - 1 in
     let flush start i =
@@ -43,7 +42,7 @@ end = struct
       end
     and escape amperstr start i =
       flush start i ;
-      add amperstr ;
+      Buffer.add_string buf amperstr ;
       let next = i + 1 in
       loop next next
     in
