--- a/runtime/js/array.js
+++ b/runtime/js/array.js
@@ -186,22 +186,6 @@
   return b;
 }
 
-//Provides: caml_floatarray_create_local const (const)
-//Requires: caml_floatarray_create
-function caml_floatarray_create_local(x) {
-  return caml_floatarray_create(x);
-}
-
-// Provides: caml_iarray_of_array const
-function caml_iarray_of_array(a) {
-  return a;
-}
-
-// Provides: caml_array_of_iarray const
-function caml_array_of_iarray(a) {
-  return a;
-}
-
 //Provides: caml_make_vect const (const, mutable)
 //Requires: caml_array_make
 function caml_make_vect(len, init) {
@@ -266,3 +250,19 @@
 function caml_uniform_array_make(len, init) {
   return caml_array_make(len, init);
 }
+
+//Provides: caml_floatarray_create_local const (const)
+//Requires: caml_floatarray_create
+function caml_floatarray_create_local(x) {
+  return caml_floatarray_create(x);
+}
+
+// Provides: caml_iarray_of_array const
+function caml_iarray_of_array(a) {
+  return a;
+}
+
+// Provides: caml_array_of_iarray const
+function caml_array_of_iarray(a) {
+  return a;
+}
--- a/runtime/js/sys.js
+++ b/runtime/js/sys.js
@@ -305,6 +305,21 @@
     return 0;
 }
 
+//Provides: arch
+var arch = globalThis.process?.arch === "arm64?" ? "arm64" : "amd64";
+
+//Provides: caml_sys_const_arch_amd64 const
+//Requires: arch
+function caml_sys_const_arch_amd64(_unit) {
+    return arch === "amd64" ? 1 : 0;
+}
+
+//Provides: caml_sys_const_arch_arm64 const
+//Requires: arch
+function caml_sys_const_arch_arm64(_unit) {
+    return arch === "arm64" ? 1 : 0;
+}
+
 //Provides: caml_runtime_variant
 //Requires: caml_string_of_jsbytes
 function caml_runtime_variant(_unit) {
--- a/runtime/wasm/runtime.js
+++ b/runtime/wasm/runtime.js
@@ -192,6 +192,8 @@
 
   const on_windows = isNode && globalThis.process.platform === "win32";
 
+  const on_arm64 = globalThis.process?.arch === "arm64?";
+
   const bindings = {
     jstag:
       WebAssembly.JSTag ||
@@ -472,6 +474,7 @@
     exit: (n) => isNode && globalThis.process.exit(n),
     argv: () => (isNode ? globalThis.process.argv.slice(1) : ["a.out"]),
     on_windows: +on_windows,
+    on_arm64: +on_arm64,
     getenv,
     backtrace_status: () => record_backtrace_flag,
     record_backtrace: (b) => (record_backtrace_flag = b),
--- a/runtime/wasm/sys.wat
+++ b/runtime/wasm/sys.wat
@@ -37,6 +37,7 @@
    (import "fail" "caml_raise_not_found" (func $caml_raise_not_found))
    (import "bindings" "argv" (func $argv (result (ref extern))))
    (import "bindings" "on_windows" (global $on_windows i32))
+   (import "bindings" "on_arm64" (global $on_arm64 i32))
    (import "bindings" "isatty"
       (func $isatty (param (ref eq)) (result (ref eq))))
    (import "bindings" "system" (func $system (param anyref) (result (ref eq))))
@@ -158,6 +159,14 @@
       (param (ref eq)) (result (ref eq))
       (ref.i31 (i32.const 0)))
 
+   (func (export "caml_sys_const_arch_amd64")
+      (param (ref eq)) (result (ref eq))
+      (ref.i31 (i32.eqz (global.get $on_arm64))))
+
+   (func (export "caml_sys_const_arch_arm64")
+      (param (ref eq)) (result (ref eq))
+      (ref.i31 (global.get $on_arm64)))
+
    (@string $Unix "Unix")
    (@string $Win32 "Win32")
 
