--- a/runtime/js/domain.js
+++ b/runtime/js/domain.js
@@ -7,6 +7,7 @@
 //Version: >= 5
 function caml_domain_dls_set(a) {
   caml_domain_dls = a;
+  return 0;
 }

 //Provides: caml_domain_dls_compare_and_set
@@ -25,6 +26,22 @@
   return caml_domain_dls;
 }

+//Provides: caml_domain_tls
+var caml_domain_tls = [0];
+
+//Provides: caml_domain_tls_set
+//Requires: caml_domain_tls
+function caml_domain_tls_set(a) {
+  caml_domain_tls = a;
+  return 0;
+}
+
+//Provides: caml_domain_tls_get
+//Requires: caml_domain_tls
+function caml_domain_tls_get(unit) {
+  return caml_domain_tls;
+}
+
 //Provides: caml_atomic_load
 //Version: >= 5
 function caml_atomic_load(ref) {
--- a/runtime/wasm/domain.wat
+++ b/runtime/wasm/domain.wat
@@ -163,5 +163,15 @@
    (func (export "caml_domain_dls_get") (param (ref eq)) (result (ref eq))
       (global.get $caml_domain_dls))

+   (global $caml_domain_tls (mut (ref eq))
+      (array.new_fixed $block 1 (ref.i31 (i32.const 0))))
+
+   (func (export "caml_domain_tls_set") (param $a (ref eq)) (result (ref eq))
+      (global.set $caml_domain_tls (local.get $a))
+      (ref.i31 (i32.const 0)))
+
+   (func (export "caml_domain_tls_get") (param (ref eq)) (result (ref eq))
+      (global.get $caml_domain_tls))
+
    (global $caml_ml_domain_unique_token (ref eq)
       (array.new_fixed $block 1 (ref.i31 (i32.const 0))))
