--- a/q.mli
+++ b/q.mli
@@ -1,3 +1,4 @@
+@@portable
 (**
    Rationals.
 
@@ -21,9 +22,9 @@
 
 (** {1 Types} *)
 
-type t = {
-    num: Z.t; (** Numerator. *)
-    den: Z.t; (** Denominator, >= 0 *)
+type t : immutable_data = {
+    num: Z.t @@ global; (** Numerator. *)
+    den: Z.t @@ global; (** Denominator, >= 0 *)
   }
 (** A rational is represented as a pair numerator/denominator, reduced to
     have a non-negative denominator and no common factor.
@@ -99,7 +100,7 @@ type kind =
     whether the numerator and/or denominator is null.
  *)
 
-val classify: t -> kind
+val classify: t @ local -> kind
 (** Determines the kind of a rational. *)
 
 val is_real: t -> bool
@@ -110,6 +111,8 @@ val sign: t -> int
     negative (including -inf), and 0 if it is null or undefined.
  *)
 
+val compare__local : t @ local -> t @ local -> int
+
 val compare: t -> t -> int
 (** [compare x y] compares [x] to [y] and returns 1 if [x] is strictly
     greater that [y], -1 if it is strictly smaller, and 0 if they are
@@ -124,6 +127,8 @@ val compare: t -> t -> int
     the ordering of rationals.
  *)
 
+val equal__local : t @ local -> t @ local -> bool
+
 val equal: t -> t -> bool
 (** Equality testing.
     Unlike [compare], this follows IEEE semantics: [undef] <> [undef].
