--- a/src-unix/notty_unix.ml
+++ b/src-unix/notty_unix.ml
@@ -35,8 +35,8 @@
 
   let set_winch_handler f =
     let signum = winch_number () in
-    let old_hdl = Sys.(signal signum (Signal_handle (fun _ -> f ()))) in
-    `Revert (once @@ fun () -> Sys.set_signal signum old_hdl)
+    let old_hdl = Sys.(Safe.signal signum (Signal_handle (fun _ -> f ()))) in
+    `Revert (once @@ fun () -> Sys.Safe.set_signal signum old_hdl)
 
   module Gen_output (O : sig
     type fd
