--- a/examples/keys.ml
+++ b/examples/keys.ml
@@ -36,6 +36,7 @@
 let pp_mouse fmt = function
   | `Release -> pps fmt "Release"
   | `Drag    -> pps fmt "Drag"
+  | `Hover   -> pps fmt "Hover"
   | `Move    -> pps fmt "Move"
   | `Press k ->
       pps fmt "Press ";
--- a/src/notty.ml
+++ b/src/notty.ml
@@ -80,7 +80,7 @@
 
     type t =
       | Ascii of string * int * int
-      | Utf8  of string * int array * int * int [@@deriving equal]
+      | Utf8  of string * int array * int * int [@@deriving equal ~localize]
   end
 
   let equal t1 t2 = match (t1, t2) with
@@ -170,9 +170,9 @@
   let replicatec w c = replicateu w (Uchar.of_char c)
 end
 
-type 'a or_unset = Unset | Set of 'a [@@deriving equal]
+type 'a or_unset = Unset | Set of 'a [@@deriving equal ~localize]
 
-let resolve_color color = 
+let resolve_color color =
   match color with
   | Unset -> 0
   | Set color -> color
@@ -183,9 +183,9 @@
   include struct
     open! Base
 
-    type color = int [@@deriving equal]
-    type style = int [@@deriving equal]
-type t = { fg : color or_unset; bg : color or_unset; st : style ; url : string option } [@@deriving equal]
+    type color = int [@@deriving equal ~localize]
+    type style = int [@@deriving equal ~localize]
+type t = { fg : color or_unset; bg : color or_unset; st : style ; url : string option } [@@deriving equal ~localize]
   end
 
   let black        = 0x01000000
@@ -204,7 +204,7 @@
   and lightmagenta = 0x0100000d
   and lightcyan    = 0x0100000e
   and lightwhite   = 0x0100000f
-  and default      = 0 
+  and default      = 0
 
   let tag c = (c land 0x03000000) lsr 24
 
@@ -487,7 +487,7 @@
     type t =
       End
     | Skip of int * t
-    | Text of A.t * Text.t * t [@@deriving equal]
+    | Text of A.t * Text.t * t [@@deriving equal ~localize]
   end
 
   let skip n k = if n = 0 then k else match k with
@@ -574,6 +574,7 @@
   ; cr      : op
   ; altscr  : bool -> op
   ; mouse   : bool -> op
+  ; hover   : bool -> op
   ; bpaste  : bool -> op
   ; cursor_nextline  : op
   ; set_title        : string -> op
@@ -589,14 +590,14 @@
 
   let sgr { A.fg; bg; st; url } ~inside buf =
     let fg = resolve_color fg in
-    let bg = resolve_color bg in 
+    let bg = resolve_color bg in
     (* NOTE: [inside] is a callback. Some attrs (e.g. the "link") attrs, need to send
        some codes at the _beginning_ of some text, and also at the _end_ of some text, so
        this `inside` callback is used to implement the things that should be drawn
        "in-between" (e.g. the visual part of a hyperlink.) *)
-    let after () = 
+    let after () =
       match url with
-      | None ->  () 
+      | None ->  ()
       | Some _ -> buf <| "\x1b]8;;\x1b\\";
     in
     ( match url with
@@ -642,7 +643,7 @@
     ; cr      = (fun b -> b <| "\x1b[1G")
     ; clreol  = (fun b -> b <| "\x1b[K")
     ; cursvis = (fun x b -> b <| if x then "\x1b[34h\x1b[?25h" else "\x1b[?25l")
-    ; cursor_kind = (fun kind b -> 
+    ; cursor_kind = (fun kind b ->
                       let n =
                         match kind with
                         | `Default -> "0"
@@ -657,6 +658,7 @@
                       b <| code)
     ; mouse   = (fun x b -> b <| if x then "\x1b[?1000;1002;1005;1015;1006h"
                                       else "\x1b[?1000;1002;1005;1015;1006l")
+    ; hover   = (fun x b -> b <| if x then "\x1b[?1003h" else "\x1b[?1003l")
     ; bpaste  = (fun x b -> b <| if x then "\x1b[?2004h" else "\x1b[?2004l")
     ; set_title = (fun title b -> b <| [%string "\027]0;%{title}\007"])
     ; save_title = (fun b -> b <| "\x1b[22;0t")
@@ -683,6 +685,7 @@
     ; cursor_kind = no1
     ; sgr     =( fun _ ~inside buffer -> inside buffer)
     ; mouse   = no1
+    ; hover   = no1
     ; bpaste  = no1
     ; set_title = no1
     ; save_title = no0
@@ -704,7 +707,7 @@
   let text_op cap buf a x = cap.sgr ~inside:(fun buf -> Text.to_buffer buf x) a buf
 
   let rec line ~screen_width ~line_width cap buf = function
-  | End              
+  | End
   | Skip (0,    End) ->
     (* NOTE: If we are at the end of a line, we do not call [erase] (which sends cleareol)
        as terminal emulators behave differently when a fully filled column receives a
@@ -721,8 +724,8 @@
   | ln::lns -> line ~screen_width ~line_width:0 cap buf ln; cap.newline buf; lines ~screen_width cap buf lns
 
 
-  let line_diffed ~screen_width ~line_width ~on_draw cap buf (prev_operation, operation) = 
-    match [%equal: Operation.t] prev_operation operation with
+  let line_diffed ~screen_width ~line_width ~on_draw cap buf (prev_operation, operation) =
+    match ([%equal: Operation.t] [@mode local]) prev_operation operation with
     | true -> cap.cursor_nextline buf
     | false -> line ~screen_width ~line_width cap buf operation; on_draw ()
   ;;
@@ -738,16 +741,16 @@
 
   let to_buffer' ~buf ~cap ~dim ~previous_operations ~operations =
     match previous_operations with
-    | None -> lines ~screen_width:(fst dim) cap buf operations 
+    | None -> lines ~screen_width:(fst dim) cap buf operations
     | Some previous_operations ->
       let is_same_height =
         List.length operations = List.length previous_operations
       in
       (match is_same_height with
-        | false -> 
-          lines ~screen_width:(fst dim) cap buf operations 
+        | false ->
+          lines ~screen_width:(fst dim) cap buf operations
         | true ->
-          lines_diffed ~screen_width:(fst dim) cap buf (Base.List.zip_exn previous_operations operations ) 
+          lines_diffed ~screen_width:(fst dim) cap buf (Base.List.zip_exn previous_operations operations )
       )
   ;;
 
@@ -796,7 +799,7 @@
 
   type key = [ special | `Uchar of Uchar.t  | `ASCII of char ] * mods
 
-  type mouse = [ `Press of button | `Drag | `Release ] * (int * int) * mods
+  type mouse = [ `Press of button | `Drag | `Hover | `Release ] * (int * int) * mods
 
   type paste = [ `Start | `End ]
 
@@ -977,7 +980,7 @@
           | ('M', (#button as b), false) -> Some (`Press b)
           | ('M', #button, true)         -> Some `Drag
           | ('m', #button, false)        -> Some `Release
-          (* | ('M', `ALL   , true)         -> Some `Move *)
+          | ('M', `ALL   , true)         -> Some `Hover
           | _                            -> None
         ) >>| fun e -> `Mouse (e, (x - 1, y - 1), mods)
 
@@ -988,8 +991,7 @@
           | (#button as b, false) -> Some (`Press b)
           | (#button     , true ) -> Some `Drag
           | (`ALL        , false) -> Some `Release
-          (* | (`ALL        , true)  -> Some `Move *)
-          | _                     -> None
+          | (`ALL        , true)  -> Some `Hover
         ) >>| fun e -> `Mouse (e, (x - 1, y - 1), mods)
 
     | CSI ("",[200],'~') -> Some (`Paste `Start)
@@ -1042,6 +1044,8 @@
   ; mutable dim   : (int * int)
   ; mutable image : I.t
   ; mutable dead  : bool
+  ; mutable mouse : bool
+  ; mutable hover : bool
   ; mutable previous_lines : Operation.t list option
   }
 
@@ -1060,23 +1064,41 @@
 
   let reset_cursor cap = cap.cursvis true & cap.cursor_kind `Default
 
-  let create ~mouse ~bpaste cap = {
+  let mouse_reporting (cap : Cap.t) ~mouse:mouse_enabled ~hover:hover_enabled =
+    match mouse_enabled, hover_enabled with
+    | false, false | false, true -> cap.hover false & cap.mouse false
+    | true, false -> cap.hover false & cap.mouse true
+    | true, true -> cap.mouse true & cap.hover true
+
+  let current_mouse_reporting t = mouse_reporting t.cap ~mouse:t.mouse ~hover:t.hover
+
+  let create ~mouse ~hover ~bpaste cap = {
       cap
     ; curs  = None
     ; dim   = (0, 0)
     ; image = I.empty
     ; dead  = false
+    ; mouse
+    ; hover
     ; write =
-        cap.altscr true & cursor cap None & cap.mouse mouse & cap.bpaste bpaste
+        cap.altscr true & cursor cap None
+        & mouse_reporting cap ~mouse ~hover
+        & cap.bpaste bpaste
     ; previous_lines = None
     }
 
-  let set_mouse t enabled = emit t (t.cap.mouse enabled)
+  let set_mouse t enabled =
+    t.mouse <- enabled;
+    emit t (current_mouse_reporting t)
+
+  let set_hover t enabled =
+    t.hover <- enabled;
+    emit t (current_mouse_reporting t)
 
   let release t =
     if t.dead then false else
       ( emit t ( t.cap.altscr false & t.cap.cursvis true &
-                 t.cap.mouse false & t.cap.bpaste false );
+                 t.cap.mouse false & t.cap.hover false & t.cap.bpaste false );
         emit t (reset_cursor t.cap);
         t.dead <- true; true )
 
@@ -1087,7 +1109,7 @@
       t.cap.sync_start &
       cursor t.cap None &
       cursat0 t.cap 0 0 &
-             (fun buf -> 
+             (fun buf ->
                 let operations = Operation.of_image (0, 0) dim image in
                 Render.to_buffer' ~buf ~cap:t.cap ~dim ~previous_operations:t.previous_lines ~operations;
                 t.previous_lines <- (Some operations);
--- a/src/notty.mli
+++ b/src/notty.mli
@@ -495,8 +495,15 @@
   type key = [ special | `Uchar of Uchar.t | `ASCII of char ] * mods
   (** Keypress event. *)
 
-  type mouse = [ `Press of button | `Drag | `Release ] * (int * int) * mods
-  (** Mouse event. *)
+  type mouse = [ `Press of button | `Drag | `Hover | `Release ] * (int * int) * mods
+  (** Mouse event.
+
+      [`Hover] events are motion events that occur while no button is depressed. They are
+      only reported when the terminal has mouse reporting and "any-event" mouse tracking
+      enabled (see {{!Tmachine.create}[Tmachine.create ~mouse ~hover]}). Note that
+      enabling this mode causes the terminal to report a mouse event for every cell the
+      pointer crosses, which can be a significant volume of input for some applications.
+  *)
 
   type paste = [ `Start | `End ]
   (** Paste event. *)
@@ -666,7 +673,7 @@
 
   type t
 
-  val create  : mouse:bool -> bpaste:bool -> Cap.t -> t
+  val create  : mouse:bool -> hover:bool -> bpaste:bool -> Cap.t -> t
   val release : t -> bool
   val output  : t -> Buffer.t -> unit
 
@@ -674,10 +681,19 @@
 
       This controls whether the terminal is asked to send mouse events on stdin.
       Disabling mouse reporting is important for allowing users to use terminal-native
-      text selection.
+      text selection. Disabling mouse reporting also disables hover reporting until mouse
+      reporting is enabled again.
   *)
   val set_mouse : t -> bool -> unit
 
+  (** Dynamically enable/disable hover (any-event) mouse reporting.
+
+      When enabled, the terminal will report mouse motion as [`Hover] events even when no
+      button is pressed, as long as ordinary mouse reporting is also enabled. Enabling
+      hover reporting can generate a large volume of input events.
+  *)
+  val set_hover : t -> bool -> unit
+
   type cursor :=
     [ `Default
     | `Bar
--- a/src-unix/notty_unix.ml
+++ b/src-unix/notty_unix.ml
@@ -162,11 +162,11 @@
         t.input.Input.cleanup ();
         write t )
 
-  let create ?(dispose=true) ?(nosig=true) ?(mouse=true) ?(bpaste=true)
+  let create ?(dispose=true) ?(nosig=true) ?(mouse=true) ?(hover=false) ?(bpaste=true)
              ?(input=Unix.stdin) ?(output=Unix.stdout) () =
     let rec t = {
         output  = Unix.out_channel_of_descr output
-      ; trm     = Tmachine.create ~mouse ~bpaste (cap_for_fd input)
+      ; trm     = Tmachine.create ~mouse ~hover ~bpaste (cap_for_fd input)
       ; buf     = Buffer.create 4096
       ; input   = Input.create ~nosig input
       ; fds     = (input, output)
--- a/src-unix/notty_unix.mli
+++ b/src-unix/notty_unix.mli
@@ -26,11 +26,12 @@
   val create : ?dispose:bool ->
                ?nosig:bool ->
                ?mouse:bool ->
+               ?hover:bool ->
                ?bpaste:bool ->
                ?input:Unix.file_descr ->
                ?output:Unix.file_descr ->
                unit -> t
-  (** [create ~dispose ~nosig ~mouse ~input ~output ()] creates a fresh
+  (** [create ~dispose ~nosig ~mouse ~hover ~input ~output ()] creates a fresh
       {{!t}terminal}. It has the following side effects:
       {ul
       {- [Unix.tcsetattr] is applied to [input] to disable {e echo} and
@@ -50,6 +51,10 @@
 
       [~mouse] activates mouse reporting. Defaults to [true].
 
+      [~hover] activates {e any-event} mouse reporting when [~mouse] is also enabled.
+      When enabled, the terminal also reports mouse motion without any button pressed as
+      [`Hover] events. Defaults to [false].
+
       [~bpaste] activates bracketed paste reporting. Defaults to [true].
 
       [~input] is the input file descriptor. Defaults to [stdin].
