--- a/runtime/js/sys.js
+++ b/runtime/js/sys.js
@@ -275,6 +275,11 @@
   return (0x7fffffff / 4) | 0;
 }
 
+//Provides: caml_sys_io_buffer_size const
+function caml_sys_io_buffer_size() {
+  return 65536;
+}
+
 //Provides: caml_sys_const_ostype_unix const
 //Requires: os_type
 function caml_sys_const_ostype_unix() {
--- a/runtime/wasm/sys.wat
+++ b/runtime/wasm/sys.wat
@@ -156,6 +156,10 @@
       (param (ref eq)) (result (ref eq))
       (ref.i31 (i32.const 0xfffffff)))
 
+   (func (export "caml_sys_io_buffer_size")
+      (param (ref eq)) (result (ref eq))
+      (ref.i31 (i32.const 65536)))
+
    (func (export "caml_sys_const_ostype_unix")
       (param (ref eq)) (result (ref eq))
       (ref.i31 (i32.eqz (global.get $on_windows))))
