--- a/runtime/js/sys.js
+++ b/runtime/js/sys.js
@@ -133,6 +133,17 @@
   return caml_string_of_jsstring(r);
 }
 
+//Provides: caml_sys_getenv_opt (const)
+//Requires: caml_string_of_jsstring
+//Requires: caml_jsstring_of_string
+//Requires: jsoo_sys_getenv
+function caml_sys_getenv_opt(name) {
+  var r = jsoo_sys_getenv(caml_jsstring_of_string(name));
+  if (r === undefined) return 0;
+  return [0, caml_string_of_jsstring(r)];
+}
+
+
 //Provides: caml_sys_unsafe_getenv
 //Requires: caml_sys_getenv
 function caml_sys_unsafe_getenv(name) {
--- a/runtime/wasm/sys.wat
+++ b/runtime/wasm/sys.wat
@@ -79,6 +79,19 @@
             (call $caml_raise_not_found)))
       (return_call $caml_string_of_jsstring (call $wrap (local.get $res))))
 
+   (func (export "caml_sys_getenv_opt")
+      (param (ref eq)) (result (ref eq))
+      (local $res anyref)
+      (local.set $res
+         (call $getenv
+            (call $unwrap (call $caml_jsstring_of_string (local.get 0)))))
+      (if (result (ref eq)) (i32.eqz (call $jsstring_test (local.get $res)))
+         (then
+            (ref.i31 (i32.const 0)))
+         (else
+            (array.new_fixed $block 2 (ref.i31 (i32.const 0))
+               (call $caml_string_of_jsstring (call $wrap (local.get $res)))))))
+
    (func (export "caml_sys_argv") (param (ref eq)) (result (ref eq))
       ;; ZZZ
       (call $caml_js_to_string_array (call $argv)))
