commit aae25c16a7f093a7c23bb966951e4c73daf8b54c
Author: Jon Ludlam <jon@recoil.org>
Date:   Fri Aug 22 09:50:54 2025 +0100

    Fix for more recent oxcamls

diff --git a/lib/top/compat_top.ml b/lib/top/compat_top.ml
index 9c6bafa3d..72bfd14bf 100644
--- a/lib/top/compat_top.ml
+++ b/lib/top/compat_top.ml
@@ -107,7 +107,9 @@ let extension_constructor
 #if OCAML_VERSION >= (5, 3, 0)
   ; ext_uid = Uid.mk ~current_unit:None
 #elif OCAML_VERSION >= (4, 11, 0)
-  ; ext_uid = Uid.mk ~current_unit:(Some (Compilation_unit.of_string "mdx"))
+  ; ext_uid = (
+    let unit_info = Unit_info.make ~check_modname:false ~source_file:"__NONE__" ~for_pack_prefix:Compilation_unit.Prefix.empty Intf "mdx" in
+    Uid.mk ~current_unit:(Some unit_info))
 #endif
   }
 
