diff --git a/src/odoc/bin/main.ml b/src/odoc/bin/main.ml
index 6f1903c..af6e86d 100644
--- a/src/odoc/bin/main.ml
+++ b/src/odoc/bin/main.ml
@@ -1,15 +1,5 @@
 #syntax quotations off
 
-(* CR-someday lmaurer: We should put ($) back into Cmdliner and then this can go away. *)
-module Cmdliner = struct
-  include Cmdliner
-
-  module Term = struct
-    include Term
-    let ($) = ($$)
-  end
-end
-
 (* CR-someday trefis: the "deps" and "targets" subcommands currently output
    their result on stdout.
    It would make the interaction with jenga nicer if we could specify a file to
diff --git a/src/search/odoc_html_frontend.ml b/src/search/odoc_html_frontend.ml
index 4f31e64..90312e0 100644
--- a/src/search/odoc_html_frontend.ml
+++ b/src/search/odoc_html_frontend.ml
@@ -19,7 +19,6 @@ end = struct
 
   let add_escape_string buf s =
     (* https://discuss.ocaml.org/t/html-encoding-of-string/4289/4 *)
-    let add = Buffer.add_string buf in
     let len = String.length s in
     let max_idx = len - 1 in
     let flush start i =
@@ -38,7 +37,7 @@ end = struct
         | _ -> loop start (i + 1)
     and escape amperstr start i =
       flush start i;
-      add amperstr;
+      Buffer.add_string buf amperstr;
       let next = i + 1 in
       loop next next
     in
diff --git a/src/syntax_highlighter/syntax_highlighter.ml b/src/syntax_highlighter/syntax_highlighter.ml
index 01bbcaf..3ea39bf 100644
--- a/src/syntax_highlighter/syntax_highlighter.ml
+++ b/src/syntax_highlighter/syntax_highlighter.ml
@@ -66,7 +66,6 @@ let tag_of_token (tok : Parser.token) =
   | INHERIT -> "INHERIT"
   | INITIALIZER -> "INITIALIZER"
   | INT _ -> "INT"
-  | KIND_ABBREV -> "KIND_ABBREV"
   | KIND_OF -> "KIND_OF"
   | LABEL _ -> "LABEL"
   | LAZY -> "LAZY"
