--- a/src/block_parser.ml
+++ b/src/block_parser.ml
@@ -330,9 +330,10 @@
       if off >= String.length s then (Buffer.contents buf, None)
       else
         match s.[off] with
-        | '\n' -> (Buffer.contents buf, Some (succ off))
+        | '\n' ->
+          (Buffer.contents buf, Some (succ off))
         | '\r' ->
-            if cr_read then Buffer.add_char buf '\r';
+          if cr_read then Buffer.add_char buf '\r';
             loop true (succ off)
         | c ->
             if cr_read then Buffer.add_char buf '\r';
@@ -345,10 +346,10 @@
     let link_defs = ref [] in
     let rec loop state = function
       | None ->
-          let blocks = finish link_defs state in
-          (blocks, List.rev !link_defs)
+        let blocks = finish link_defs state in
+        (blocks, List.rev !link_defs)
       | Some off ->
-          let s, off = read_line s off in
+        let s, off = read_line s off in
           loop (process link_defs state s) off
     in
     loop empty (Some 0)
--- a/src/parser.ml
+++ b/src/parser.ml
@@ -23,6 +23,10 @@
 (** Raised when a parser fails, used for control flow rather than error
     handling *)
 
+exception Max_recursion_depth of string
+(** Raised when a function reaches the maximum recursion depth specified
+    (namely, in emphasis detection).  *)
+
 (** Stateful parser combinators *)
 module P : sig
   type state
@@ -966,9 +970,16 @@
 
   let rec parse_emph = function
     | (Emph (pre, _, q1, n1) as x1) :: xs when is_opener x1 ->
-        let rec loop acc = function
-          | (Emph (_, post, q2, n2) as x2) :: xs1 as xs
-            when is_closer x2 && q1 = q2 ->
+      let rec loop acc emphs depth =
+
+        if depth = 0 then 
+          raise (Max_recursion_depth 
+                   "Max recursion depth reached on emphasis detection. \
+                    Should your text actually be code?");
+        let depth = depth - 1 in
+        match emphs with
+        | (Emph (_, post, q2, n2) as x2) :: xs1 as xs
+          when is_closer x2 && q1 = q2 ->
               (* At this point we have an openener followed by a closer. Both are of the same style (either * or _) *)
               if (is_opener x2 || is_closer x1) && not (is_emph_match n1 n2)
               then
@@ -987,9 +998,10 @@
                  *)
                 match find_next_emph xs1 with
                 | Some (_, _, _, n3) when is_emph_match n3 n2 ->
-                    let xs' = parse_emph xs in
-                    loop acc xs'
-                | _ -> loop (x2 :: acc) xs1
+                  let xs' = parse_emph xs in
+                    loop acc xs' depth
+                | _ ->
+                  loop (x2 :: acc) xs1 depth
               else
                 let xs =
                   if n1 >= 2 && n2 >= 2 then
@@ -1010,7 +1022,7 @@
                   else r
                 in
                 parse_emph r
-          | (Emph (_, _, q2, _) as x2) :: xs1 as xs when is_opener x2 ->
+        | (Emph (_, _, q2, _) as x2) :: xs1 as xs when is_opener x2 ->
               (*
                This case happens when we encounter a second opener delimiter. We look ahead for the next closer,
                and if the next closer is of the same style, we can match them together.
@@ -1029,12 +1041,12 @@
                 | None -> false
                 | Some (_, _, q3, _) -> q2 = q3
               in
-              if not is_next_closer_same then loop (x2 :: acc) xs1
-              else loop acc (parse_emph xs)
-          | x :: xs -> loop (x :: acc) xs
+              if not is_next_closer_same then loop (x2 :: acc) xs1 depth
+              else loop acc (parse_emph xs) depth
+          | x :: xs -> loop (x :: acc) xs depth
           | [] -> x1 :: List.rev acc
         in
-        loop [] xs
+        loop [] xs 10000  (* Default max depth of 10000 *)
     | x :: xs -> x :: parse_emph xs
     | [] -> []
 
@@ -1883,7 +1895,8 @@
         junk st;
         Buffer.add_char buf c;
         loop ~seen_link acc st
-    | exception Fail -> Pre.parse_emph (List.rev (text acc))
+    | exception Fail ->
+      Pre.parse_emph (List.rev (text acc))
   in
   loop ~seen_link:false [] st
 
