--- a/src/notty.ml
+++ b/src/notty.ml
@@ -539,7 +539,7 @@
 
   let ansi = {
       skip    = (fun n b -> b <| "\x1b[0m"; Buffer.add_chars b ' ' n)
-    ; newline = (fun b -> b <| "\x1bE")
+    ; newline = (fun b -> b <| "\x1b[E")
     ; altscr  = (fun x b -> b <| if x then "\x1b[?1049h" else "\x1b[?1049l")
     ; cursat  = (fun w h b -> b <| "\x1b["; b <! h; b <. ';'; b <! w; b <. 'H')
     ; cubcuf  = (fun x b -> b <| "\x1b["; b <! abs x; b <. if x < 0 then 'D' else 'C')
@@ -586,7 +586,6 @@
   let rec line cap buf = function
     End              -> erase cap buf
   | Skip (n,    End) -> erase cap buf; skip_op cap buf n
-  | Text (a, x, End) -> erase cap buf; text_op cap buf a x
   | Skip (n,    ops) -> skip_op cap buf n; line cap buf ops
   | Text (a, x, ops) -> text_op cap buf a x; line cap buf ops
 
