--- a/src/notty.ml
+++ b/src/notty.ml
@@ -494,7 +494,8 @@
   let (&) op1 op2 buf = op1 buf; op2 buf
 
   type cursor =
-    [ `Bar
+    [ `Default
+    | `Bar
     | `Bar_blinking
     | `Block
     | `Block_blinking
@@ -560,6 +561,7 @@
     ; cursor_kind = (fun kind b -> 
                       let n =
                         match kind with
+                        | `Default -> "0"
                         | `Bar -> "6"
                         | `Bar_blinking -> "5"
                         | `Block -> "2"
@@ -894,6 +896,8 @@
     | None        -> cap.cursvis false
     | Some (w, h, cursor_kind) -> cap.cursvis true & cursat0 cap w h & cap.cursor_kind cursor_kind
 
+  let reset_cursor cap = cap.cursvis true & cap.cursor_kind `Default
+
   let create ~mouse ~bpaste cap = {
       cap
     ; curs  = None
@@ -908,6 +912,7 @@
     if t.dead then false else
       ( emit t ( t.cap.altscr false & t.cap.cursvis true &
                  t.cap.mouse false & t.cap.bpaste false );
+        emit t (reset_cursor t.cap);
         t.dead <- true; true )
 
   let output t buf = t.write buf; t.write <- ignore
--- a/src/notty.mli
+++ b/src/notty.mli
@@ -622,7 +622,8 @@
   val output  : t -> Buffer.t -> unit
 
   type cursor :=
-    [ `Bar
+    [ `Default
+    | `Bar
     | `Bar_blinking
     | `Block
     | `Block_blinking
--- a/src-unix/notty_unix.mli
+++ b/src-unix/notty_unix.mli
@@ -74,7 +74,8 @@
       Useful if the output might have become garbled. *)
 
   type cursor :=
-    [ `Bar
+    [ `Default
+    | `Bar
     | `Bar_blinking
     | `Block
     | `Block_blinking
