diff --git a/lib/write.ml b/lib/write.ml
index 1234567..abcdefg 100644
--- a/lib/write.ml
+++ b/lib/write.ml
@@ -201,9 +201,9 @@ let write_std_float ob x;
   Buffer.contents ob


-let write_intlit = Buffer.add_string
-let write_floatlit = Buffer.add_string
-let write_stringlit = Buffer.add_string
+let write_intlit ob s = Buffer.add_string ob s
+let write_floatlit ob s = Buffer.add_string ob s
+let write_stringlit ob s = Buffer.add_string ob s

 let rec iter2_aux f_elt f_sep x = function
     [] -> ()
