--- b/compiler/lib-wasm/generate.ml
+++ a/compiler/lib-wasm/generate.ml
@@ -111,23 +111,32 @@
       ; "caml_int32_compare", (`Pure, [ Int32; Int32 ], Int Normalized)
       ; "caml_nativeint_compare", (`Pure, [ Nativeint; Nativeint ], Int Normalized)
       ; "caml_int64_compare", (`Pure, [ Int64; Int64 ], Int Normalized)
+      ; "caml_string_geti8", (`Mutator, [ Value; Int Normalized ], Int Normalized)
+      ; "caml_string_geti16", (`Mutator, [ Value; Int Normalized ], Int Normalized)
       ; "caml_string_get16", (`Mutator, [ Value; Int Normalized ], Int Normalized)
       ; "caml_string_get32", (`Mutator, [ Value; Int Normalized ], Int32)
       ; "caml_string_get64", (`Mutator, [ Value; Int Normalized ], Int64)
       ; "caml_string_getf32", (`Mutator, [ Value; Int Normalized ], Float32)
+      ; "caml_bytes_geti8", (`Mutator, [ Value; Int Normalized ], Int Normalized)
+      ; "caml_bytes_geti16", (`Mutator, [ Value; Int Normalized ], Int Normalized)
       ; "caml_bytes_get16", (`Mutator, [ Value; Int Normalized ], Int Normalized)
       ; "caml_bytes_get32", (`Mutator, [ Value; Int Normalized ], Int32)
       ; "caml_bytes_get64", (`Mutator, [ Value; Int Normalized ], Int64)
       ; "caml_bytes_getf32", (`Mutator, [ Value; Int Normalized ], Float32)
+      ; "caml_bytes_set8", (`Mutator, [ Value; Int Normalized; Int Unnormalized ], Value)
       ; "caml_bytes_set16", (`Mutator, [ Value; Int Normalized; Int Unnormalized ], Value)
       ; "caml_bytes_set32", (`Mutator, [ Value; Int Normalized; Int32 ], Value)
       ; "caml_bytes_set64", (`Mutator, [ Value; Int Normalized; Int64 ], Value)
       ; "caml_bytes_setf32", (`Mutator, [ Value; Int Normalized; Float32 ], Value)
       ; "caml_lxm_next", (`Mutable, [ Value ], Int64)
+      ; "caml_ba_uint8_geti8", (`Mutator, [ Value; Int Normalized ], Int Normalized)
+      ; "caml_ba_uint8_geti16", (`Mutator, [ Value; Int Normalized ], Int Normalized)
       ; "caml_ba_uint8_get16", (`Mutator, [ Value; Int Normalized ], Int Normalized)
       ; "caml_ba_uint8_get32", (`Mutator, [ Value; Int Normalized ], Int32)
       ; "caml_ba_uint8_get64", (`Mutator, [ Value; Int Normalized ], Int64)
       ; "caml_ba_uint8_getf32", (`Mutator, [ Value; Int Normalized ], Float32)
+      ; ( "caml_ba_uint8_set8"
+        , (`Mutator, [ Value; Int Normalized; Int Unnormalized ], Value) )
       ; ( "caml_ba_uint8_set16"
         , (`Mutator, [ Value; Int Normalized; Int Unnormalized ], Value) )
       ; "caml_ba_uint8_set32", (`Mutator, [ Value; Int Normalized; Int32 ], Value)
--- b/compiler/lib/specialize_js.ml
+++ a/compiler/lib/specialize_js.ml
@@ -301,14 +301,20 @@
 let getters =
   indexing_primitives
     [ "caml_array_get"
+    ; "caml_string_geti8"
+    ; "caml_string_geti16"
     ; "caml_string_get16"
     ; "caml_string_get32"
     ; "caml_string_get64"
     ; "caml_string_getf32"
+    ; "caml_bytes_geti8"
+    ; "caml_bytes_geti16"
     ; "caml_bytes_get16"
     ; "caml_bytes_get32"
     ; "caml_bytes_get64"
     ; "caml_bytes_getf32"
+    ; "caml_ba_uint8_geti8"
+    ; "caml_ba_uint8_geti16"
     ; "caml_ba_uint8_get16"
     ; "caml_ba_uint8_get32"
     ; "caml_ba_uint8_get64"
@@ -318,10 +324,12 @@
 let setters =
   indexing_primitives
     [ "caml_array_set"
+    ; "caml_bytes_set8"
     ; "caml_bytes_set16"
     ; "caml_bytes_set32"
     ; "caml_bytes_set64"
     ; "caml_bytes_setf32"
+    ; "caml_ba_uint8_set8"
     ; "caml_ba_uint8_set16"
     ; "caml_ba_uint8_set32"
     ; "caml_ba_uint8_set64"
--- b/runtime/js/bigarray.js
+++ a/runtime/js/bigarray.js
@@ -548,6 +548,20 @@
   return ba.get(ofs);
 }
 
+//Provides: caml_ba_uint8_geti8
+//Requires: caml_array_bound_error
+function caml_ba_uint8_geti8(ba, i0) {
+  var ofs = ba.offset(i0);
+  if (ofs >= ba.data.length) caml_array_bound_error();
+  return (ba.get(ofs) << 24) >> 24;
+}
+
+//Provides: caml_ba_uint8_geti16
+//Requires: caml_ba_uint8_get16
+function caml_ba_uint8_geti16(ba, i0) {
+  return (caml_ba_uint8_get16(ba, i0) << 16) >> 16;
+}
+
 //Provides: caml_ba_uint8_get16
 //Requires: caml_array_bound_error
 function caml_ba_uint8_get16(ba, i0) {
@@ -611,6 +625,15 @@
   return 0;
 }
 
+//Provides: caml_ba_uint8_set8
+//Requires: caml_array_bound_error
+function caml_ba_uint8_set8(ba, i0, v) {
+  var ofs = ba.offset(i0);
+  if (ofs >= ba.data.length) caml_array_bound_error();
+  ba.set(ofs + 0, v & 0xff);
+  return 0;
+}
+
 //Provides: caml_ba_uint8_set16
 //Requires: caml_array_bound_error
 function caml_ba_uint8_set16(ba, i0, v) {
--- b/runtime/js/mlBytes.js
+++ a/runtime/js/mlBytes.js
@@ -141,6 +141,33 @@
   return caml_string_unsafe_get(s, i);
 }
 
+//Provides: caml_string_geti8
+//Requires: caml_string_unsafe_get, caml_string_bound_error
+//Requires: caml_ml_string_length
+function caml_string_geti8(s, i) {
+  if (i >>> 0 >= caml_ml_string_length(s)) caml_string_bound_error();
+  return (caml_string_unsafe_get(s, i) << 24) >> 24;
+}
+
+//Provides: caml_bytes_geti8
+//Requires: caml_bytes_unsafe_get, caml_bytes_bound_error
+function caml_bytes_geti8(s, i) {
+  if (i >>> 0 >= s.l) caml_bytes_bound_error();
+  return (caml_bytes_unsafe_get(s, i) << 24) >> 24;
+}
+
+//Provides: caml_string_geti16
+//Requires: caml_string_get16
+function caml_string_geti16(s, i) {
+  return (caml_string_get16(s, i) << 16) >> 16;
+}
+
+//Provides: caml_bytes_geti16
+//Requires: caml_bytes_get16
+function caml_bytes_geti16(s, i) {
+  return (caml_bytes_get16(s, i) << 16) >> 16;
+}
+
 //Provides: caml_string_get16
 //Requires: caml_string_unsafe_get, caml_string_bound_error
 //Requires: caml_ml_string_length
@@ -230,6 +257,15 @@
   return caml_string_unsafe_set(s, i, c);
 }
 
+//Provides: caml_bytes_set8
+//Requires: caml_bytes_bound_error, caml_bytes_unsafe_set
+function caml_bytes_set8(s, i, i8) {
+  if (i >>> 0 >= s.l) caml_bytes_bound_error();
+  var b1 = 0xff & i8;
+  caml_bytes_unsafe_set(s, i + 0, b1);
+  return 0;
+}
+
 //Provides: caml_bytes_set16
 //Requires: caml_bytes_bound_error, caml_bytes_unsafe_set
 function caml_bytes_set16(s, i, i16) {
--- b/runtime/wasm/bigarray.wat
+++ a/runtime/wasm/bigarray.wat
@@ -2215,6 +2215,25 @@
                (br $loop))))
       (return (i32.const 0)))
 
+   (func (export "caml_ba_uint8_geti8")
+      (param $vba (ref eq)) (param $i i32) (result i32)
+      (local $ba (ref $bigarray))
+      (local $view (ref extern))
+      (local.set $ba (ref.cast (ref $bigarray) (local.get $vba)))
+      (local.set $view (struct.get $bigarray $ba_view (local.get $ba)))
+      (if (i32.lt_s (local.get $i) (i32.const 0))
+         (then (call $caml_bound_error)))
+      (if (i32.ge_u (local.get $i)
+             (array.get $int_array
+                (struct.get $bigarray $ba_dim (local.get $ba))
+                (i32.const 0)))
+         (then (call $caml_bound_error)))
+      (i32.extend8_s (call $dv_get_ui8 (local.get $view) (local.get $i))))
+
+   (func (export "caml_ba_uint8_geti16")
+      (param $vba (ref eq)) (param $i i32) (result i32)
+      (i32.extend16_s (call $caml_ba_uint8_get16 (local.get $vba) (local.get $i))))
+
    (func (export "caml_ba_uint8_get16")
       (param $vba (ref eq)) (param $i i32) (result i32)
       (local $ba (ref $bigarray))
@@ -2263,6 +2282,23 @@
       (call $dv_get_i64_unaligned
          (local.get $view) (local.get $i) (i32.const 1)))
 
+   (func (export "caml_ba_uint8_set8")
+      (param $vba (ref eq)) (param $i i32) (param $d i32)
+      (result (ref eq))
+      (local $ba (ref $bigarray))
+      (local $view (ref extern))
+      (local.set $ba (ref.cast (ref $bigarray) (local.get $vba)))
+      (local.set $view (struct.get $bigarray $ba_view (local.get $ba)))
+      (if (i32.lt_s (local.get $i) (i32.const 0))
+         (then (call $caml_bound_error)))
+      (if (i32.ge_u (local.get $i)
+             (array.get $int_array
+                (struct.get $bigarray $ba_dim (local.get $ba))
+                (i32.const 0)))
+         (then (call $caml_bound_error)))
+      (call $dv_set_i8 (local.get $view) (local.get $i) (local.get $d))
+      (ref.i31 (i32.const 0)))
+
    (func (export "caml_ba_uint8_set16")
       (param $vba (ref eq)) (param $i i32) (param $d i32)
       (result (ref eq))
--- b/runtime/wasm/string.wat
+++ a/runtime/wasm/string.wat
@@ -152,6 +152,23 @@
          (i31.get_u (ref.cast (ref i31) (local.get $len))))
       (ref.i31 (i32.const 0)))
 
+   (export "caml_string_geti8" (func $caml_bytes_geti8))
+   (func $caml_bytes_geti8 (export "caml_bytes_geti8")
+      (param $v (ref eq)) (param $p i32) (result i32)
+      (local $s (ref $bytes))
+      (local.set $s (ref.cast (ref $bytes) (local.get $v)))
+      (if (i32.lt_s (local.get $p) (i32.const 0))
+         (then (call $caml_bound_error)))
+      (if (i32.ge_u (local.get $p)
+                    (array.len (local.get $s)))
+         (then (call $caml_bound_error)))
+      (i32.extend8_s (array.get_u $bytes (local.get $s) (local.get $p))))
+
+   (export "caml_string_geti16" (func $caml_bytes_geti16))
+   (func $caml_bytes_geti16 (export "caml_bytes_geti16")
+      (param $v (ref eq)) (param $p i32) (result i32)
+      (i32.extend16_s (call $caml_bytes_get16 (local.get $v) (local.get $p))))
+
    (export "caml_string_get16" (func $caml_bytes_get16))
    (func $caml_bytes_get16 (export "caml_bytes_get16")
       (param $v (ref eq)) (param $p i32) (result i32)
@@ -240,6 +257,18 @@
                               (i32.add (local.get $p) (i32.const 7))))
                         (i64.const 56))))))
 
+   (func (export "caml_bytes_set8")
+      (param (ref eq)) (param $p i32) (param $v i32) (result (ref eq))
+      (local $s (ref $bytes))
+      (local.set $s (ref.cast (ref $bytes) (local.get 0)))
+      (if (i32.lt_s (local.get $p) (i32.const 0))
+         (then (call $caml_bound_error)))
+      (if (i32.ge_u (local.get $p)
+                    (array.len (local.get $s)))
+         (then (call $caml_bound_error)))
+      (array.set $bytes (local.get $s) (local.get $p) (local.get $v))
+      (ref.i31 (i32.const 0)))
+
    (func (export "caml_bytes_set16")
       (param (ref eq)) (param $p i32) (param $v i32) (result (ref eq))
       (local $s (ref $bytes))
