--- a/runtime/js/domain.js
+++ b/runtime/js/domain.js
@@ -126,6 +126,12 @@
   return 1;
 }
 
+//Provides: caml_max_domain_count
+//Version: >= 5
+function caml_max_domain_count(unit) {
+  return 1;
+}
+
 //Provides: caml_ml_domain_index
 //Requires: caml_domain_id
 //Version: >= 5.03
--- a/runtime/wasm/domain.wat
+++ b/runtime/wasm/domain.wat
@@ -178,6 +178,10 @@
       (param (ref eq)) (result (ref eq))
       (ref.i31 (i32.const 1)))
 
+   (func (export "caml_max_domain_count")
+      (param (ref eq)) (result (ref eq))
+      (ref.i31 (i32.const 1)))
+
    (global $caml_domain_id (export "caml_domain_id") (mut i32) (i32.const 0))
    (global $caml_domain_latest_id (export "caml_domain_latest_id") (mut i32)
       (i32.const 1))
