--- a/compiler/lib/magic_number.ml
+++ b/compiler/lib/magic_number.ml
@@ -64,27 +64,11 @@
 
 let equal a b = compare a b = 0
 
-let v =
-  let current = Ocaml_version.current in
-  match current with
-  | 4 :: 13 :: _ -> 30
-  | 4 :: 14 :: _ -> 31
-  | 5 :: 00 :: _ -> 32
-  | 5 :: 01 :: _ -> 33
-  | 5 :: 02 :: _ -> 34
-  | 5 :: 03 :: _ -> 35
-  | _ ->
-      if Ocaml_version.compare current [ 4; 13 ] < 0
-      then failwith "OCaml version unsupported. Upgrade to OCaml 4.13 or newer."
-      else (
-        assert (Ocaml_version.compare current [ 5; 4 ] >= 0);
-        failwith "OCaml version unsupported. Upgrade js_of_ocaml.")
+let current_exe = of_string Ocaml_common.Config.exec_magic_number
 
-let current_exe = "Caml1999X", v
+let current_cmo = of_string Ocaml_common.Config.cmo_magic_number
 
-let current_cmo = "Caml1999O", v
-
-let current_cma = "Caml1999A", v
+let current_cma = of_string Ocaml_common.Config.cma_magic_number
 
 let current = function
   | `Exe -> current_exe
